jib-core/src/main/java/com/google/cloud/tools/jib/http/FailoverHttpClient.java

```diff
@@ -223,32 +243,49 @@ public Response call(String httpMethod, URL url, Request request) throws IOExcep
 
       try {
         logInsecureHttpsFailover(url);
-        return call(httpMethod, url, request, getHttpTransport(false));
+        Response response = call(httpMethod, url, request, getHttpTransport(false));
+        failoverHistory.put(url.getHost() + ":" + url.getPort(), Failover.INSECURE_HTTPS);
+        return response;
 
       } catch (SSLException ignored) { // This is usually when the server is plain-HTTP.
         logHttpFailover(url);
-        return call(httpMethod, toHttp(url), request, getHttpTransport(true));
+        Response response = call(httpMethod, toHttp(url), request, getHttpTransport(true));
+        failoverHistory.put(url.getHost() + ":" + url.getPort(), Failover.HTTP);
+        return response;
       }
 
     } catch (ConnectException ex) {
       // It is observed that Open/Oracle JDKs sometimes throw SocketTimeoutException but other times
       // ConnectException for connection timeout. (Could be a JDK bug.) Note SocketTimeoutException
       // does not extend ConnectException (or vice versa), and we want to be consistent to error out
       // on timeouts: https://github.com/GoogleContainerTools/jib/issues/1895#issuecomment-527544094
-      if (ex.getMessage() != null && ex.getMessage().contains("timed out")) {
-        throw ex;
-      }
-
-      // Fall back to HTTP only if "url" had no port specified (i.e., we tried the default HTTPS
-      // port 443) and we could not connect to 443. It's worth trying port 80.
-      if (enableHttpAndInsecureFailover && isHttpsProtocol(url) && url.getPort() == -1) {
-        logHttpFailover(url);
-        return call(httpMethod, toHttp(url), request, getHttpTransport(true));
+      if (ex.getMessage() == null || !ex.getMessage().contains("timed out")) {
+        // Fall back to HTTP only if "url" had no port specified (i.e., we tried the default HTTPS
+        // port 443) and we could not connect to 443. It's worth trying port 80.
+        if (enableHttpAndInsecureFailover && isHttpsProtocol(url) && url.getPort() == -1) {
+          logHttpFailover(url);
+          Response response = call(httpMethod, toHttp(url), request, getHttpTransport(true));
+          failoverHistory.put(url.getHost() + ":" + url.getPort(), Failover.HTTP);
+          return response;
+        }
       }
       throw ex;
     }
   }
 
+  private Optional<Response> followFailoverHistory(String httpMethod, URL url, Request request)
+      throws IOException {
+    Preconditions.checkArgument(isHttpsProtocol(url));
+    switch (failoverHistory.getOrDefault(url.getHost() + ":" + url.getPort(), Failover.NONE)) {
+      case INSECURE_HTTPS:
```

Comment:
should we add a comment that this order is the reverse of how we failover? Incase someone the in the future tries to add something here, but has no context?

Reply:
There are two orders actually.

1. secure HTTPS --> insecure HTTPS --> plain HTTP
1. secure HTTPS --> plain HTTP: no HTTPS attempt in this case.

But the order of the `case`s in the `switch` doesn't matter. How exactly do you want to comment this?

Reply:
Re-ordered the cases anyways. Added a comment too.

Reply:
oh you know what, I wasn't even thinking that the map would only hold **one** failover value, I'm dumb. You should order this in whatever way makes logical sense to you.